contracts/Registry.sol

```diff
@@ -1,70 +1,652 @@
+        // delete the tags associated with the service
+        for (uint i = orgsByName[orgName].servicesByName[serviceName].tags.length; i > 0; i--) {
+            removeTagFromServiceRegistration(orgName, serviceName, orgsByName[orgName].servicesByName[serviceName].tags[i-1]);
+        }
+
+        // swap lut entries
+        uint    indexToUpdate   = orgsByName[orgName].servicesByName[serviceName].orgServiceIndex;
+        bytes32 serviceToUpdate = orgsByName[orgName].serviceKeys[orgsByName[orgName].serviceKeys.length-1];
+
+        if (orgsByName[orgName].serviceKeys[indexToUpdate] != serviceToUpdate) {
+            orgsByName[orgName].serviceKeys[indexToUpdate] = serviceToUpdate;
+            orgsByName[orgName].servicesByName[serviceToUpdate].orgServiceIndex = indexToUpdate;
+        }
+
+        orgsByName[orgName].serviceKeys.length--;
+
+        // delete contents of service registration
+        delete orgsByName[orgName].servicesByName[serviceName];
+    }
+
+    //   _____                        ____                        __  __                 _
+    //  |_   _|   _ _ __   ___       |  _ \ ___ _ __   ___       |  \/  | __ _ _ __ ___ | |_
+    //    | || | | | '_ \ / _ \      | |_) / _ \ '_ \ / _ \      | |\/| |/ _` | '_ ` _ \| __|
+    //    | || |_| | |_) |  __/      |  _ <  __/ |_) | (_) |     | |  | | (_| | | | | | | |_
+    //    |_| \__, | .__/ \___|      |_| \_\___| .__/ \___/      |_|  |_|\__, |_| |_| |_|\__|
+    //        |___/|_|                         |_|                       |___/
+
+    function createTypeRepositoryRegistration(bytes32 orgName, bytes32 repositoryName, bytes32 repositoryPath,
+            bytes repositoryURI, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireTypeRepositoryExistenceConstraint(orgName, repositoryName, false);
+
+        TypeRepositoryRegistration memory typeRepo;
+        orgsByName[orgName].typeReposByName[repositoryName] = typeRepo;
+        orgsByName[orgName].typeReposByName[repositoryName].repositoryName = repositoryName;
+        orgsByName[orgName].typeReposByName[repositoryName].repositoryPath = repositoryPath;
+        orgsByName[orgName].typeReposByName[repositoryName].repositoryURI = repositoryURI;
+        orgsByName[orgName].typeReposByName[repositoryName].orgTypeRepoIndex = orgsByName[orgName].typeRepoKeys.length;
+        orgsByName[orgName].typeRepoKeys.push(repositoryName);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToTypeRepositoryRegistration(orgName, repositoryName, tags[i]);
+        }
     }
 
-    AgentRecord[] public agentRecords;
-    mapping(bytes32 => uint) public agentIndex;
+    function updateTypeRepositoryRegistration(bytes32 orgName, bytes32 repositoryName, bytes32 repositoryPath,
+        bytes repositoryURI) external {
 
-    event RecordCreated(bytes32 name, address agent);
```

Comment:
Event cost is low; there isn't a good reason why we would omit them from this version of the Registry.